Framework/Geometry/test/DetectorInfoIteratorTest.h

```diff
@@ -0,0 +1,172 @@
+#ifndef MANTID_GEOMETRY_DETECTORINFOITERATORTEST_H_
+#define MANTID_GEOMETRY_DETECTORINFOITERATORTEST_H_
+
+#include "MantidBeamline/DetectorInfo.h"
+#include "MantidGeometry/Instrument/Detector.h"
+#include "MantidGeometry/Instrument/DetectorInfo.h"
+#include "MantidGeometry/Instrument/DetectorInfoItem.h"
+#include "MantidGeometry/Instrument/DetectorInfoIterator.h"
+#include "MantidGeometry/Instrument/InstrumentVisitor.h"
+#include "MantidGeometry/Instrument/ParameterMap.h"
+#include "MantidKernel/EigenConversionHelpers.h"
+#include "MantidKernel/V3D.h"
+#include "MantidTestHelpers/ComponentCreationHelper.h"
+
+#include <algorithm>
+#include <boost/make_shared.hpp>
+#include <cxxtest/TestSuite.h>
+#include <set>
+
+using Mantid::detid_t;
+using Mantid::Kernel::V3D;
+using namespace ComponentCreationHelper;
+using namespace Mantid::Geometry;
+
+class DetectorInfoIteratorTest : public CxxTest::TestSuite {
+public:
+  // This pair of boilerplate methods prevent the suite being created statically
+  // This means the constructor isn't called when running other tests
+  static DetectorInfoIteratorTest *createSuite() {
+    return new DetectorInfoIteratorTest();
+  }
+  static void destroySuite(DetectorInfoIteratorTest *suite) { delete suite; }
+
+  DetectorInfoIteratorTest(){};
+
+  Mantid::Geometry::DetectorInfo create_detector_info_object() {
+
+    // Create a very basic instrument to visit
+    auto visitee = createMinimalInstrument(V3D(0, 0, 0),   // Source position
+                                           V3D(10, 0, 0),  // Sample position
+                                           V3D(11, 0, 0)); // Detector position
+
+    // Add 10 more detectors to the instrument
+    for (int i = 2; i < 12; i++) {
+      Detector *det =
+          new Detector("point-detector", i /*detector id*/, nullptr);
+      det->setPos(V3D(10 + i, 0, 0));
+      std::string num = std::to_string(i);
+      det->setShape(createSphere(0.01 /*1cm*/, V3D(0, 0, 0), num));
+      visitee->add(det);
+      visitee->markAsDetector(det);
+    }
+
+    // Create the instrument visitor
+    InstrumentVisitor visitor(visitee);
+
+    // Visit everything
+    visitor.walkInstrument();
+
+    // Create the Beamline DetectorInfo
+    auto detInfo = visitor.detectorInfo();
```

Comment:
You could save a lot of lines here using the automatic wrapper creation in the `InstrumentVisitor`